Mathlib/Algebra/Algebra/NonUnitalHom.lean

```diff
@@ -79,184 +97,192 @@ namespace NonUnitalAlgHomClass
 
 -- Porting note: Made following instance non-dangerous through [...] -> [...] replacement
 -- See note [lower instance priority]
-instance (priority := 100) toNonUnitalRingHomClass {F R A B : Type*}
-    {_ : Monoid R} {_ : NonUnitalNonAssocSemiring A} [DistribMulAction R A]
-    {_ : NonUnitalNonAssocSemiring B} [DistribMulAction R B] [FunLike F A B]
-    [NonUnitalAlgHomClass F R A B] : NonUnitalRingHomClass F A B :=
-  { ‹NonUnitalAlgHomClass F R A B› with }
+instance (priority := 100) toNonUnitalRingHomClass
+  {F R S : Type*} [Monoid R] [Monoid S] {φ : outParam (R →* S)} {A B : Type*}
+    [NonUnitalNonAssocSemiring A] [DistribMulAction R A]
+    [NonUnitalNonAssocSemiring B] [DistribMulAction S B] [FunLike F A B]
+    [NonUnitalAlgSemiHomClass F φ A B] : NonUnitalRingHomClass F A B :=
+  { ‹NonUnitalAlgSemiHomClass F φ A B› with } -- coe := (⇑) }
 #align non_unital_alg_hom_class.non_unital_alg_hom_class.to_non_unital_ring_hom_class NonUnitalAlgHomClass.toNonUnitalRingHomClass
 
 variable [Semiring R] [NonUnitalNonAssocSemiring A] [Module R A]
-  [NonUnitalNonAssocSemiring B] [Module R B]
+  [Semiring S] [NonUnitalNonAssocSemiring B] [Module S B]
+  {φ : R →+* S}
 
+-- Why do I need to specify R and S ?
 -- see Note [lower instance priority]
-instance (priority := 100) {F R A B : Type*}
-    {_ : Semiring R} {_ : NonUnitalSemiring A} {_ : NonUnitalSemiring B}
-    [Module R A] [Module R B]
-    [FunLike F A B] [NonUnitalAlgHomClass F R A B] :
-    LinearMapClass F R A B :=
-  { ‹NonUnitalAlgHomClass F R A B› with map_smulₛₗ := map_smul }
-
-/-- Turn an element of a type `F` satisfying `NonUnitalAlgHomClass F R A B` into an actual
-`NonUnitalAlgHom`. This is declared as the default coercion from `F` to `A →ₙₐ[R] B`. -/
-@[coe]
-def toNonUnitalAlgHom {F R A B : Type*} [Monoid R] [NonUnitalNonAssocSemiring A]
-    [DistribMulAction R A] [NonUnitalNonAssocSemiring B] [DistribMulAction R B] [FunLike F A B]
-    [NonUnitalAlgHomClass F R A B] (f : F) : A →ₙₐ[R] B :=
-  { (f : A →ₙ+* B) with
-    map_smul' := map_smul f }
+instance (priority := 100) {F : Type*} [FunLike F A B]
+    [NonUnitalAlgSemiHomClass (R := R) (S := S) F φ A B] :
+    SemilinearMapClass F φ A B :=
+  { ‹NonUnitalAlgSemiHomClass F φ A B› with map_smulₛₗ := map_smulₛₗ }
 
-instance {F R A B : Type*} [Monoid R] [NonUnitalNonAssocSemiring A] [DistribMulAction R A]
+-- TODO (ACL) : Why is this needed? Adjust priority?
+instance (priority := 100) {F : Type*} [FunLike F A B] [Module R B] [NonUnitalAlgHomClass F R A B] :
+    LinearMapClass F R A B :=
+  { ‹NonUnitalAlgHomClass F R A B› with map_smulₛₗ := map_smulₛₗ }
```

Comment:
Priority 100 looks good to me.

The reason we need this instance is because the `SemilinearMapClass` instance is about `↑(φ : RingHom R S) : MonoidHom R S` and Lean doesn't see that `↑RingHom.id = MonoidHom.id`. (Much like the problem with `DistribMulActionHomClass` we saw below. The easiest resolution is to keep a pair of instances for the semi case and the non-semi case. Alternatively, all the classes need to be more generic in what they accept: either turning all these arguments into plain functions `R → S`. Or perhaps we can use classes here too...